### PR TITLE
Remove deploys in finalized blocks

### PR DESCRIPTION
**Pull request by KentShikama**
### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>

https://rchain.atlassian.net/browse/RHOL-1127

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala [hunk 1]
```diff
@@ -142,11 +142,25 @@ class MultiParentCasperImpl[F[_]: Sync: Concurrent: Capture: ConnectionsCell: Tr
...
+                              )
                             case None => lastFinalizedBlockHash.pure[F]
                           }
     } yield newFinalizedBlock
 
+  private def removeDeploysInFinalizedBlock(finalizedChild: BlockHash): F[Unit] =
+    for {
+      b              <- ProtoUtil.unsafeGetBlock[F](finalizedChild)
+      deploys        = b.body.get.deploys.map(_.deploy.get)
+      deploysRemoved = deploys.count(deployHist.remove)
```
Comment (itegulov) [inline casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala]:
```suggestion
      deploysRemoved = deploys.toList.foldM(0)((count, deploy) =>
        Monad[F].ifM(Sync[F].delay { deployHist.remove(deploy) })((count + 1).pure[F], count.pure[F])
      )
```
Or something similar as `deployHist.remove` is effectful

**Comment by KentShikama** [comment]
bors try

**Comment by KentShikama** [comment]
bors try

File: casper/src/test/scala/coop/rchain/casper/HashSetCasperTest.scala [hunk 2]
```diff
@@ -1336,19 +1338,20 @@ class HashSetCasperTest extends FlatSpec with Matchers {
...
 
       createBlock8Result <- nodes(1).casperEff
                              .deploy(deployDatas(7)) *> nodes(1).casperEff.createBlock
       Created(block8) = createBlock8Result
-      _               = nodes(1).casperEff.addBlock(block8, ignoreDoppelgangerCheck[Effect])
-      _               = nodes(0).receive()
-      _               = nodes(2).receive()
+      _               <- nodes(1).casperEff.addBlock(block8, ignoreDoppelgangerCheck[Effect])
+      _               <- nodes(0).receive()
+      _               <- nodes(2).receive()
```
Comment (itegulov) [inline casper/src/test/scala/coop/rchain/casper/HashSetCasperTest.scala]: Sorry, seems like I have introduced this bug while I was refactoring casper tests

**Pull request merged**